handle invalid key names

erothman-groupon opened the pull request:
1. Check for invalid key names, such as hyphens, special characters, names that begin with integers, etc.
2. Provide a valid name and map to the original name via a resolve function.

This solves the validation error problem when running graphiql UI, for instance:
"Names must match /^[_a-zA-Z][_a-zA-Z0-9]*$/ but \"1\" does not."
"Names must match /^[_a-zA-Z][_a-zA-Z0-9]*$/ but \"Send-Id\" does not."

Here is an example of the generated field name for a key that has an invalid hyphen:
```javascript
fields: () => ({
    sendId: {
      type: GraphQLString,
      resolve: obj => obj['Send-Id']
    }
  })
```
Here is an example for a key named with an integer:
```javascript
fields: () => ({
    _1: {
      type: GraphQLInt,
      resolve: obj => obj['1']
    }
})
```
Further reading and a good example is at: https://graphql.org/graphql-js/type/#examples
```javascript
var AddressType = new GraphQLObjectType({
  name: 'Address',
  fields: {
    street: { type: GraphQLString },
    number: { type: GraphQLInt },
    formatted: {
      type: GraphQLString,
      resolve(obj) {
        return obj.number + ' ' + obj.street
      }
    }
  }
});
```
To handle mutations (post and put requests), the generator updates the resolver method to handle special character fields. In the following example, three fields are updated to the original source name before being posted back to the server:
 ```javascript
updatePet: {
      type: UpdatePetResponse,
      args: {
        body: {
          type: GraphQLNonNull(PetInput)
        },
        debug: {
          type: GraphQLString
        }
      },

      async resolve(parent, options, context, info) {
        if (context[VERIFY_AUTH_STATUS]) {
          const securityConfig = {
            "security": [{
              "petstore_auth": ["write:pets", "read:pets"]
            }],
            "definitions": {
              "petstore_auth": {
                "type": "oauth2",
                "authorizationUrl": "https://petstore.swagger.io/oauth/dialog",
                "flow": "implicit",
                "scopes": {
                  "write:pets": "modify pets in your account",
                  "read:pets": "read your pets"
                }
              }
            }
          };
          context[VERIFY_AUTH_STATUS](securityConfig);
        }

        const mapFields = [{
          name: 'isNickName',
          originalName: 'is-nick-name',
          paramType: 'body'
        }, {
          name: 'ageInDogYears',
          originalName: 'age-in-dog-years',
          paramType: 'body'
        }, {
          name: '_5ThingsAreNeato',
          originalName: '5 Things are Neato!',
          paramType: 'body'
        }];

        if (mapFields && mapFields.length) {
          mapFields.forEach(({
            name,
            originalName,
            paramType
          }) => {
            const value = options[paramType][name];
            options[paramType][originalName] = value;
            delete options[paramType][name];
          });
        }

        const data = await context[FETCH]('/pet', {
          method: 'PUT',
          endpointName: 'updatePet',
          qs: {
            'debug': options.debug
          },
          pathParams: undefined,
          headers: undefined,
          json: options.body
        }).text();
        return {
          rawResponseBody: data,
          rawInputOptions: options
        };
      }

    }
```

dbushong commented:
See: https://github.com/groupon/swagql/blob/master/CONTRIBUTING.md#commits--commit-messages for notes about commit message requirements

dbushong commented:
`resolve` is a good find!  I think I'd like to make the property names as pretty as possible, since those are the ones the developers making and consuming queries are going to see - hence skip the leading `_` if we can avoid it, and maybe default to camelCase to keep things short?

I realize it complicates things due to the need to then track potential dups, but we only get to do this once as a non-breaking change.

dbushong commented on lib/generator/type-map.js:
```diff
@@ -93,21 +98,47 @@ function generateTypeAST(typeMap, type) {
+          ),
+        ])
+      );
+    }
```
This is a big enough chunk that it could be in an informatively named function, or at least have a comment - babel AST code is super-hard to read once it's been more than a few weeks since you've touched it, so I lean heavily on anything we can do to make it more readable.

erothman-groupon commented:
Note I'm still not sure how to handle key names with integers. I'm open to suggestions! It's a head-scratcher.

dbushong commented on lib/generator/type-map.js:
```diff
@@ -92,23 +96,14 @@ function generateTypeAST(typeMap, type) {
   }
 
   const requiredProps = apiType.required || [];
+  const usedNames = Object.keys(props);
```
```suggestion
  const usedNames = new Set(Object.keys(props));
```

dbushong commented on lib/generator/type-map.js:
```diff
@@ -138,6 +133,60 @@ function generateTypeAST(typeMap, type) {
+      : camelCase(originalName);
+    let uniqIdx = 2;
+    let finalName = newName;
+    while (includes(usedNames, finalName)) {
```
```suggestion
    while (usedNames.has(finalName)) {
```

dbushong commented on lib/generator/type-map.js:
```diff
@@ -138,6 +133,60 @@ function generateTypeAST(typeMap, type) {
+    const idx = findIndex(usedNames, k => {
+      return k === originalName;
+    });
+    usedNames[idx] = finalName;
```
this could then just be something like:
```js
usedNames.delete(originalName);
usedNames.add(finalName);
```

dbushong commented on lib/generator/type-map.js:
```diff
@@ -138,6 +133,60 @@ function generateTypeAST(typeMap, type) {
+  function makeValidName(originalName) {
+    const newName = /^[0-9]/.test(originalName)
+      ? `_${originalName}`
+      : camelCase(originalName);
```
I think you want something like:
```js
const newName = camelCase(originalName.replace(/^(?=\d)/, '_'));
```
so that if you have something like: '5 Things Are - Neato!' it'll still end up `_5thingsAreNeato` or whatever

dbushong commented:
Coming along well; thanks!  Since I've never played with the `resolve` feature before, I'd like to see it tested in practice with a client request.  I think we could add a bogus property to one of the petstore swagger json fixtures and then somewhere around here:
https://github.com/groupon/swagql/blob/0638d85432a0eb53bee5f05812b8550c2c37348d/test/generator/swagql.test.js#L103
we could verify that the right data came thru.

dbushong commented on lib/generator/type-map.js:
```diff
@@ -138,6 +132,69 @@ function generateTypeAST(typeMap, type) {
+    return t.objectProperty(
+      t.identifier(newName),
+      t.objectExpression([
+        t.objectProperty(t.identifier('type'), typeMap.ref({ type: 'string' })),
```
How do we know this is always type string?

dbushong commented on test/generator/swagql.test.js:
```diff
@@ -12,14 +12,23 @@ const http = require('http');
       }
     });
-    await server.listen(8080);
+    await server.listen(port);
```
I'm not sure how this ever worked... vanilla node http server.listen doesn't return a Promise, AFAIK

I'd just make line 15 `let port;` and then here we can do:

```js
port = await new Promise(resolve => {
  server.listen(0, () => {
    resolve(server.address().port); 
  });
});
```

dbushong commented on lib/generator/type-map.js:
```diff
@@ -92,23 +95,14 @@ function generateTypeAST(typeMap, type) {
-        ),
-      ])
-    );
+    return validName.test(propKey)
```
should this be `validName.test(propKey) || isInput` and then we can skip the conditional branching later?

dbushong commented on lib/generator/type-map.js:
```diff
@@ -138,6 +132,71 @@ function generateTypeAST(typeMap, type) {
+            requiredProps.includes(propKey),
+            type,
+            propKey
+          )
```
can we DRY some of this type clause stuff?

dbushong commented:
Looking good - so this works recursively too, right?  if the response is `{ safe: { alsoSafe: { 'Totally Not Safe!': { '4': 2 } } } }` it still works?

dbushong commented on lib/generator/type-map.js:
```diff
@@ -138,6 +130,51 @@ function generateTypeAST(typeMap, type) {
+        )
+      );
+    }
+    const name = isValidName ? propKey : makeValidName(propKey);
```
so if we're potentially remapping the name for input fields and not providing a resolver, how will that work?

dbushong commented:
Can you briefly walk me through what the most recent commit does?  ("create map fields resolve...")

erothman-groupon commented:
The goal was to create a resolve function that could update the special character fields back to their original values. The original code actually does this for certain types of parameters like query and header params (lines 232-273) but does not handle invalid names for body params (line 267). It sort of just passes them through. The post and put operations use the body param so this PR focuses on those http operations only (line 285). It loops through the body parameters and if it finds a param with invalid characters, it creates an AST object with the following properties: 1) a valid name, 2) the original source name, and 3) the param type, which in this case is always 'body' (lines 296 to 306). The object is pushed to the mapFields array. Then on line 336, the MAP_FIELDS is added to the resolve options. These resolve options are passed to the template functions. There are different template functions for the different operations. I created a new template function called buildMapFieldsResolveFunction on line 87. It's basically the same template that was used previously for post and put operations but with the added MAP_FIELDS array injected and a method to update the field names to their original names (lines 97-103) before the data is sent back to the endpoint. On line 346, the script checks to see if we need to use the mapped fields template based on whether it's a post or put operation and whether it has invalid character fields. Otherwise it uses the original resolver template. The result in the schema file for the updated petstore swagger file:

```javascript
updatePet: {
      type: UpdatePetResponse,
      args: {
        body: {
          type: GraphQLNonNull(PetInput)
        },
        debug: {
          type: GraphQLString
        }
      },

      async resolve(parent, options, context, info) {
        if (context[VERIFY_AUTH_STATUS]) {
          const securityConfig = {
            "security": [{
              "petstore_auth": ["write:pets", "read:pets"]
            }],
            "definitions": {
              "petstore_auth": {
                "type": "oauth2",
                "authorizationUrl": "https://petstore.swagger.io/oauth/dialog",
                "flow": "implicit",
                "scopes": {
                  "write:pets": "modify pets in your account",
                  "read:pets": "read your pets"
                }
              }
            }
          };
          context[VERIFY_AUTH_STATUS](securityConfig);
        }

        const mapFields = [{
          name: 'isNickName',
          originalName: 'is-nick-name',
          paramType: 'body'
        }, {
          name: 'ageInDogYears',
          originalName: 'age-in-dog-years',
          paramType: 'body'
        }, {
          name: '_5ThingsAreNeato',
          originalName: '5 Things are Neato!',
          paramType: 'body'
        }];

        if (mapFields && mapFields.length) {
          mapFields.forEach(({
            name,
            originalName,
            paramType
          }) => {
            const value = options[paramType][name];
            options[paramType][originalName] = value;
            delete options[paramType][name];
          });
        }

        const data = await context[FETCH]('/pet', {
          method: 'PUT',
          endpointName: 'updatePet',
          qs: {
            'debug': options.debug
          },
          pathParams: undefined,
          headers: undefined,
          json: options.body
        }).text();
        return {
          rawResponseBody: data,
          rawInputOptions: options
        };
      }

    }
```

dbushong commented on lib/generator/deoperationalize.js:
```diff
@@ -81,6 +84,31 @@ async function resolve(parent, options, context, info) {
+    context[VERIFY_AUTH_STATUS](securityConfig);
+  }
+
+  const mapFields = MAP_FIELDS;
```
should this be `${MAP_FIELDS}`?

erothman-groupon replied:
I thought so too originally but that throws an error. It's actually like the variable defined here: https://github.com/groupon/swagql/blob/d388f8a265636feef8a1a7992ffad032f75db109/lib/generator/deoperationalize.js#L134

dbushong commented on lib/generator/deoperationalize.js:
```diff
@@ -81,6 +84,31 @@ async function resolve(parent, options, context, info) {
+      options[paramType][originalName] = value;
+      delete options[paramType][name];
+    });
+  }
```
does this handle nested bad property names?

erothman-groupon replied:
i added a PR to handle nested property objects. unfortunately, it made the code a lot more complicated.

erothman-groupon commented:
To handle nested property fields, I added a path to the mapFields as such: 
```javascript
const mapFields = [{
  name: 'isNickName',
  originalName: 'is-nick-name',
  path: 'body'
}, {
  name: 'ageInDogYears',
  originalName: 'age-in-dog-years',
  path: 'body'
}, {
  name: '_5ThingsAreNeato',
  originalName: '5 Things are Neato!',
  path: 'body'
}, {
  name: 'totallyNotSafe',
  originalName: 'Totally Not Safe!',
  path: 'body.safe.alsoSafe'
}, {
  name: '_4',
  originalName: '4',
  path: 'body.safe.alsoSafe.totallyNotSafe'
}];
```
The function to rename the fields to their original values is where the complicated bit starts. I couldn't use lodash's get method so had to create my own. The fields are updated in reverse order because we need to handle the children first before deleting their parent.
```javascript
function getValue(object, keys) {
  return keys.split('.').reduce(function (o, k) {
      return (o || {})[k];
  }, object);
}

if (mapFields && mapFields.length) {
  mapFields.reverse().forEach(({
    name,
    originalName,
    path
  }) => {
    let obj = getValue(options, path);
    if (obj) {
      const value = obj[name];
      obj[originalName] = value;
      delete obj[name];
    }   
  });
}
```
Here's a hypothetical options arg that is passed to the create a new pet resolver function:
```javascript
options:  {
  "body": {
    "name": "Fido",
    "isNickName": false,
    "ageInDogYears": 5,
    "safe": {
      "alsoSafe": {
        "totallyNotSafe": {
          "_4": "nested objects work!"
        }
      }
    },
    "photoUrls": [
      "/photo.png"
    ]
  }
}
```
After the mapFields are updated, the options have the original field names and will be posted to the endpoint as such:
```javascript
options:  {
  "body": {
    "name": "Fido",
    "safe": {
      "alsoSafe": {
        "Totally Not Safe!": {
          "4": "nested objects work!"
        }
      }
    },
    "photoUrls": [
      "/photo.png"
    ],
    "age-in-dog-years": 5,
    "is-nick-name": false
  }
}
```
